keps/sig-scheduling/5278-nominated-node-name-for-expectation/README.md

```diff
@@ -0,0 +1,856 @@
+
+It helps the scenarios where the binding cycles take time, for example, VolumeBinding plugin takes time at PreBind extension point.
+
+#### Story 2: Cluster Autoscaler specifies `NominatedNodeName` to indicate where pods can go after new nodes are created/registered
+
+When the cluster autoscaler creates instances for pending pods, it calculate which new node might get which pending pod.
+If they can put `NominatedNodeName` based on those calculation, it could prevent the double effort, 
+the scheduler calculates the scheduling result at scheduling retries, scanning all the nodes in the cluster again, when those new nodes are registered at Kube-apiserver.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Race condition
+
+If an external component adds `NominatedNodeName` to the pod that is going through a scheduling cycle,
+`NominatedNodeName` isn't taken into account (of course), and the pod could be scheduled onto a different node.
+
+But, this should be fine because, either way, we're not saying `NominatedNodeName` is something force the scheduler to pick up the node,
+rather it's just a preference.
+
+#### Confusion if `NominatedNodeName` is different from `NodeName` after all
+
+If an external component adds `NominatedNodeName`, but the scheduler picks up a different node,
+`NominatedNodeName` is just overwritten by a final decision of the scheduler.
+
+But, if an external component updates `NominatedNodeName` that is set by the scheduler, 
+the pod could end up having different `NominatedNodeName` and `NodeName`.
+
+Probably we should clear `NominatedNodeName` when the pod is bound. (at binding api)
+
+#### What if there are multiple components that set `NominatedNodeName`
+
+First of all, probably we should not recommend such situation.
+
+But, one potential way to mitigate it would be having a validation rule that doesn't allow `NominatedNodeName` to be updated from one to another,
+but only allow it to be set when it's empty.
+The scheduler clears `NominatedNodeName` when the pod is unschedulable (i.e., the nominated node isn't schedulable too).
+And, after that, external components can put a new preference on `NominatedNodeName`.
+
+Another potential way to completely solve it would be adding a new API `NominatedNodeNames` on the Pod so that it can accomodate all preferences from different components. But, that would be an implementation cost and I'm not sure if we have a use case where different components want to set `NominatedNodeNames` at the same pod, or one component wants to set multiple `NominatedNodeNames`.
```

Comment:
What do you think @dom4ha @wojtek-t

Reply:
Yeah, as implied on the KEP, I'm also -1 on NominatedNodeName **_s_** idea.

But, what do you think about this `What if there are multiple components that could set `NominatedNodeName` on the same pod` topic, in general? I'm concerned that multiple controllers keep overwriting NNN that is set by the others. Of course, we can regard such situation just as user's fault though. Should we just leave it, just recommending not to have such multiple controllers on the doc or somewhere? Or do we need some safe guard to prevent such?

Reply:
I doubt that we will suddenly observe a spike in new components trying to utilize this field. So until we observe it as something problematic, I would probably limit to documenting this risk?